### DISCUSSION TITLE
How to force skipping cache in a query?

### DISCUSSION THREAD

**Discussion by Kaltsoon** [TanStack/query#2398] [Q&A] [answered]
I'm aware that using `cacheTime: 0` with `useQuery` will skip cache _if_ query is not already mounted. That being said, how to skip cache in component `B` in this example: https://codesandbox.io/s/dreamy-hoover-ejosw?file=/src/index.js. With "skip cache" in this context I mean not using the cached data once the `B` component mounts (like it would work with `cacheTime: 0` if the query wouldn't be mounted elsewhere). In my opinion, there should be a bulletproof way to skip cache in such a scenario.

To add a bit more context: we are working with forms and we need to be sure that the data is fresh once the component mounts, so we want to go through the `{ data: null, isLoading: true, ... }` => `{ data: ..., isLoading: false, ... }` transition every time the component mounts.  We have been successful so far with `cacheTime: 0`, but I noticed that cache is not ignored if query is already mounted in some other component. This is a bit of a pitfall in my opinion.

**Comment by TkDodo** [comment, marked as answer]
You’d need to provide a unique query key to achieve that.

**Reply by Kaltsoon** [reply]
But that's quite a hack, don't you think? I would need to generate some random id when the component mounts which is not related to the query at all just to make the caching work.

**Reply by TkDodo** [reply]
No, I don't think that's a hack. It looks like you want to _bypass_ the query cache. Another option would be to not use react-query for this call, because you don't want the cache - which is what react-query is good at :)

Put another way: You want a separate cached entry for every mounted instance of your component, and since react-query uses the query key to give you a previously fetched and cached entry, giving it a new query key achieves that.

**Reply by Kaltsoon** [reply]
For example if I have a such reusable hook:

```
const useUser = (id, options) => {
  const queryKey = ['user', id]

  const queryFn = () => ...

  return useQuery(queryKey, queryFn, options)
}
```

I would need to change the implementation like this:

```
const useUser = (id, options = {}) => {
  const { someKeyToSkipCache } = options
  const queryKey = ['user', id, someKeyToSkipCache]

  const queryFn = () => ...

  return useQuery(queryKey, queryFn, options)
}
```

**Reply by Kaltsoon** [reply]
I could create a decorator hook for `useQuery` which has a more explicit and bulletproof `skipCache` option, but I think it is something that I and the community would like to be implemented by `react-query`.

**Reply by TkDodo** [reply]
what alternative are you suggesting? Like a `skipCache: true` option?

I would implement it like so:

```
const useUserWithoutCache = (id, options) => {
  const uniqueId = React.useState(() => uuid.v4())[1]
  const queryKey = ['user', id, uniqueId]

  const queryFn = () => ...

  return useQuery(queryKey, queryFn, options)
}
```

so it becomes an implementation detail 🤷

**Reply by Kaltsoon** [reply]
Yeah, that's pretty much the implementation I was thinking about. I was hoping that something like `skipCache: true` option would exist in `react-query` because it is a quite common use case. E.g. Apollo Client has `fetchPolicy: 'network-only'` which I have used many times in the past.

**Reply by TkDodo** [reply]
> I'm aware that using cacheTime: 0 with useQuery will skip cache if query is not already mounted.

So to this not entirely true. What react-query does is give you any data that it has in the cache that matches the query key. With `cacheTime: 0`, you clear (garbage collect) the cached entry as soon as all observers unmount. So if you then mount again, there is nothing in the cache, which is why you get a hard loading state. So technically, the cache is not "skipped", because there is no way to "skip" the cache. If there is something for the query key, you'll get it, and that is on purpose. Having a cache is one of the prime reasons to use this library.

>  but I think it is something that I and the community would like to be implemented by react-query.

I don't fully understand the use-case. A resource is a resource. If you have a user with id 5 and you have data in your cache, why would you _not_ want it? react-query will make a background refetch anyhow, so the stale data will be updated after that. Again, this is the purpose of the lib - the whole stale-while-revalidate mechanism.

You can also look into `queryClient.removeQueries` to manually delete the cache, in which case there will again be nothing to deliver.

**Reply by TkDodo** [reply]
> `fetchPolicy: 'network-only'`

will this not fetch on every render then? I've never used apollo 😅

**Reply by TkDodo** [reply]
> we are working with forms and we need to be sure that the data is fresh once the component mounts

I'm just assuming now: Do you take the data from react-query and put it _into_ local state for your form?

**Reply by Kaltsoon** [reply]
Apollo's `fetchPolicy: 'cache-and-network'` is pretty much how react-query works. `fetchPolicy: 'network-only'` means the "cachleless" mechanism which totally bybasses the cache and fetches on the mount and when some query dependant argument changes.

**Reply by Kaltsoon** [reply]
> I'm just assuming now: Do you take the data from react-query and put it _into_ local state for your form?

Exactly, e.g. with Formik, we define the form's initial values based on the query's data. We could update those initial values when fresh data arrives from the query but that's a bit inconvenient for the users

**Reply by TkDodo** [reply]
If possible, I would try to keep server state and client state separate. I’ve written about that pattern here: https://tkdodo.eu/blog/practical-react-query#keep-server-and-client-state-separate

**Reply by Kaltsoon** [reply]
Your example is spot on what we are doing. But "Background updates are very unlikely to yield something new, and even if, your Form has already been initialized" is exactly the issue if we do not do manual cache updates after mutations. We should need to worry about these cache updates if we could disable cache in scenarios like yours.

**Reply by TkDodo** [reply]
Formik also has the `enableReinitialze` prop which would work nicely with react-query's background updates: https://github.com/formium/formik/issues/1033#issuecomment-432006334

**Reply by TkDodo** [reply]
If background updates won’t likely yield anything new, why not just use what you have in the cache 🙃

**Reply by Kaltsoon** [reply]
> Formik also has the `enableReinitialze` prop which would work nicely with react-query's background updates: [formium/formik#1033 (comment)](https://github.com/formium/formik/issues/1033#issuecomment-432006334)

Yes, but I think it is not good UX that the form's content randomly changes after the initialization at some point in time.

**Reply by Kaltsoon** [reply]
> If background updates won’t likely yield anything new, why not just use what you have in the cache 🙃

But it does after mutations as I mentioned earlier.

**Reply by Kaltsoon** [reply]
I think what I'm trying to say is that react-query's SWR does the job 90% of cases, but I don't understand why it is so insane to ask for a way to skip cache. It isn't an anti-pattern so I don't understand why there isn't an explicit and easy way to do it without custom implementations.

**Reply by TkDodo** [reply]
> Yes, but I think it is not good UX that the form's content randomly changes after the initialization at some point in time.

valid point, but imo so is showing a loading spinner instead of the form even though you have data in the cache already.

> But it does after mutations as I mentioned earlier.

you can remove the entry from the cache after a mutation with `queryClient.removeQueries`, I think that would serve the purpose here.

>  I don't understand why it is so insane to ask for a way to skip cache

It's not insane, but off the top of my head, I would see a couple of issues with such a flag.
- What happens if the component re-renders? I guess we don't want to bypass then, so it's more like: bypass the cache if a new observer mounts for the first time.
- What happens if you have multiple observers? Usually, they are all informed about the state a query is in. Now if I have one mounted observer that is in `success` state, and then one observer mounts with `skipCache: true`, would the other observer also go to `loading` state? If not, would we have two observers of the same query, where each is in a different state (one is `loading`, on is `success`)?
- What about the `onSuccess` callback - is it called for all observers then?

**Reply by Kaltsoon** [reply]
Those are valid points perhaps I was oversimplifying the `skipCache` option a bit. The solutions you proposed are quite reasonable. Thanks for the discussion!
